### PR TITLE
`General`: Increase interactivity of active students line charts and course exercises average bar chart

### PR DESCRIPTION
**Pull request by willergeorg**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR increases the interactivity for several charts:

- The line chart on the course detail page
- The line chart in Course administration
- The bar chart displaying the average student score of each exercise in course statistics

### Description

It is now possible to perform some more interaction with these chats:

- Next to the title of the line charts, a Tooltip is now placed that explains the term "active students"
- The title of the line charts are now clickable
- The line chart on the course detail page is now displayable in full screen mode
- By clicking on a bar, the user is delegated to the corresponding exercise statistics page

Besides that, some minor adjustments in the diagrams are included in this PR as well:

- The lines in the line chart are now smoothed in order to present a trend-line.
- The size difference within the tooltips is now a bit reduced
- The line chart on the course detail page now displays one additional week in the view
- The tooltips for specific points in the line chart now contain the relative value represented by the point as number
- The line chart on the course detail page contains the average over the displayed time frame as reference line that is toggleable
- For the bar chart on the course exercise, now a first color coding is introduced: If the average score is higher than 90% the bar is displayed in green, it it is below 25% it is displayed in red
- The tooltips of the bar chart now contain the exercise type of the corresponding exercise

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with some exercises and participation

1. Log in to Artemis
2. Navigate to Course Administration
3. Inspect the line chart there. Have a look at the tooltips in the chart and the explanation tooltip next to the title. Also check the translation of the explanation tooltip.
4. Click on the chart title. You should be delegated to the corresponding course detail page.
5. Inspect the more detailed line chart there. Again, check the tooltips and the explanation translation. Make sure that the reference line disappears if the check box is unchecked.
6. Also inspect the appearance of the chart in full screen mode.

-  **Important**: There is a remaining problem with the fullscreen mode of the chart: The tooltips are not visible in full screen. This is due to the underlying mechanics of `ngx-charts` (at least as far as I can tell). I found [a PR](https://github.com/swimlane/ngx-charts/pull/1114) in the GitHub project of the framework that seems to tackle this issue. But as this PR is from Freburary of 2019 at it is still open, I do not have much hope that this is still maintained. 

7. Click on the chart title again. You should be delegated to the course statistics.
8. Check the first bar chart there. Make sure that the color coding works as expected. Also check the tooltips if the exercise type is matching and the translations are appropriate/ aligned with translations at oder places in Artemis.
9. Click on a bar of your choice. You should be delegated to the corresponding exercise statistics page.

                       

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

statistics-average-score-graph.component.ts: 72.72% | 93.87%
course-detail-line-chart.component.ts: 68.42% |  92.85% 
### Screenshots

The line chart on course administration:
image

The line chart on course detail page:
image

The bar chart in the course statistics:
image

### CODE REVIEW

File: src/main/webapp/app/shared/statistics-graph/statistics-average-score-graph.component.ts [hunk 1]
```diff
@@ -40,20 +48,102 @@ export class StatisticsAverageScoreGraphComponent implements OnInit {
...
+     * @param value average score of the exercise
+     * @private
+     */
+    private determineChartEntry(name: string, value: number): ExerciseStatisticsEntry | undefined {
+        let counter = 0;
+        let result;
+        /**
+         * this is not a good solution as neither the name of the exercise nor its average score is necessarily unique.
+         * But as the exercise type is not included in the emitted event, I do not see a better solution.
+         */
```
Comment (b-fein) [inline src/main/webapp/app/shared/statistics-graph/statistics-average-score-graph.component.ts:120]:
```suggestion
        /*
         * The emitted event only contains the name and the average score of the exercise. Using those values to determine the chart entry
         * is not an ideal solution as this pair is not necessarily unique.
         * In practice they most likely are unique, though. Not being able to find the entry in this edge case therefore has negligible impact.
         */
```
I think not handling this special edge case is okay.

**Review by b-fein — DISMISSED** [review]
Code review: I added a few small inline comments, looks good otherwise.

**Review by Malyuk-A — APPROVED** [review]
Code changes (especially after the resolved change requests) look good.
Tested on TS3. (did not notice any (new) console errors) Works just as described 👍

**Review by cremertim — APPROVED** [review]
tested on ts 1, did not encounter any issues (works according to the testing steps)

**Review by b-fein — APPROVED** [review]
Code re-approve. :+1:

**Review by filip-gregurevic — APPROVED** [review]
Nice job! 👍🏼

**Review by MertAksehirlioglu — COMMENTED** [review]
Tested on ts1 and found a small issue about the translations. In the course statistics page if you change the language, everything except the days&months in charts change.

The days and months are still in German even if i change the language to English. If i refresh the page they are also translated to English.

Screen Shot 2022-01-08 at 22 25 02

**Comment by willergeorg** [comment]
> Tested on ts1 and found a small issue about the translations. In the course statistics page if you change the language, everything except the days&months in charts change.
> 
> The days and months are still in German even if i change the language to English. If i refresh the page they are also translated to English.
> 
> Screen Shot 2022-01-08 at 22 25 02
Yes, this is due to the realization of translations for these charts and I'm aware of this inconsistency. I have an idea on how this can be solved. In this PR, only the first bar chart is changed, the ones below are not affected by my changes, therefore I would prefer to deal with this in another PR. But thanks for reporting the observation. 👍

**Pull request merged**